### PR TITLE
[expo-gl-cpp] Rewrite of expo-gl-cpp

### PR DESCRIPTION
**Pull request by wkozyra95**
# Why

This is a merge of all PRs that were created during work on the rewrite. After a discussion with @nikki93 we decided to create one big PR for all the changes. Main reasons for that were
- some changes added in one PR are reverted in next one(especially typed array stuff), it might complicate review process
- some PRs change structure in a significant way that makes rebasing impossible, all review comments would need to apply on last(or separate PR)
- every PR compiled, but I didn't run extensive tests on each of them, so it might not work(the last one should be well tested)
- almost every line was affected, so it's easier to go together through the entire codebase of the module

#### Partial PRs can be found here:
- https://github.com/expo/expo/pull/6750
- https://github.com/expo/expo/pull/6982
- https://github.com/expo/expo/pull/6991
- https://github.com/expo/expo/pull/7169

# How

There are still some problems with the current implementation
- Content of `JSCRuntime::isArrayBuffer`, `JSCRuntime::data`, `JSCRuntime::isArrayBuffer` needs to be uncommented, PR to react-native is [here](https://github.com/facebook/react-native/pull/28634). There is some delay with signing CLA, so it's blocked for now, but it should be merge without any problems.
- it breaks if versioned code is attempting to call it, `expo-gl-cpp` is compiled against one jsi, but if it is used with react-native with jsi that is not ABI compatible it will fail during linking. In current implementation `expo-gl-cpp` would work only for SDKs that have the same react-native version as the latest one. 

### Merging

I would prefer to merge this PR as separate commits (without squashing), if we do this that way when TypedArray API will be merged into hermes we can just revert [this](https://github.com/expo/expo/pull/7948/commits/02f4e2fab896fbf68bd83f980ebd26e9322bf1f4) commit `add jsi compatible TypedArray API`

# Test Plan

android binaries and ios `pod install` results are on branch `@wkozyra95/migrate-expo-gl-to-jsi-build`  [here](https://github.com/expo/expo/tree/%40wkozyra95/migrate-expo-gl-to-jsi-build)

- android should work out of the box
- on iOS to make it work you would need apply the same changes in react-native-lab as the ones in `android/ReactCommon/jsi/JSCRuntime.cpp `

Tested webgl conformance test and NCL as `UNVERSIONED` project

### CODE REVIEW

File: packages/expo-gl-cpp/android/src/main/jni/Application.mk [hunk 1]
```diff
@@ -1,15 +1,14 @@
 APP_BUILD_SCRIPT := Android.mk
 
-APP_ABI := armeabi-v7a x86 arm64-v8a x86_64
+APP_ABI := arm64-v8a # armeabi-v7a x86 arm64-v8a x86_64 # TODO(wkozyra95) uncomment
```
[the remark this answers is not in the document]
Comment (wkozyra95) [reply packages/expo-gl-cpp/android/src/main/jni/Application.mk:3]: it's uncommented in PR on top of this one (with versioning)

File: packages/expo-gl-cpp/android/src/main/java/expo/modules/gl/cpp/EXGL.java [hunk 2]
```diff
@@ -7,8 +7,7 @@
   static {
     SoLoader.loadLibrary("expo-gl");
   }
-
-  public static native int EXGLContextCreate(long jsCtxPtr); // react-native < 0.59
```
[the remark this answers is not in the document]
Comment (wkozyra95) [reply packages/expo-gl-cpp/android/src/main/java/expo/modules/gl/cpp/EXGL.java:11]: At this point we can just get rid of EXGLContextCreateV2 and get back to using EXGLContextCreate. current code of expo-gl-cpp works only with react-native 0.63

**Comment by jarvisluong** [comment]
Does this mean now on master branch, expo-gl should work with Hermes engine?

**Comment by wkozyra95** [comment]
@jarvisluong

Yes, it requires react-native 0.63. I didn't test hermes after merging it to master, but it worked in the past(so it should still work).

**Pull request merged**